toonz/sources/toonz/sceneviewerevents.cpp

```diff
@@ -790,19 +787,25 @@ void SceneViewer::onRelease(const TMouseEvent &event) {
   tool->setViewer(this);
 
   {
-    TPointD pos = tool->getMatrix().inv() *
-                  winToWorld(event.mousePos() * getDevPixRatio());
-
-    TObjectHandle *objHandle = TApp::instance()->getCurrentObject();
-    if (tool->getToolType() & TTool::LevelTool && !objHandle->isSpline()) {
-      pos.x /= m_dpiScale.x;
-      pos.y /= m_dpiScale.y;
-    }
+    QPointF winPos = event.mousePos() * getDevPixRatio();
+    TPointD screenPos( winPos.x() - (double)width()/2.0,
+                      -winPos.y() + (double)height()/2.0 );
+    TPointD worldPos = getInputManager()->screenToWorld() * screenPos;
+    TPointD pos = getInputManager()->worldToTool() * worldPos;
 
     if (m_mouseButton == Qt::LeftButton || m_tabletState == Released) {
-      if (!m_toolSwitched) tool->leftButtonUp(pos, event);
+      if (!m_toolSwitched) {
+        getInputManager()->trackEvent(
+          0, 0, screenPos,
+          (event.m_isTablet ? &event.m_pressure : NULL),
+          NULL, true, TToolTimer::ticks() );
+        getInputManager()->processTracks();
+      }
```

Comment:
I figured out the "tails"/"catching up" issue and tracked it to this spot.

The problem is the pointer position on `TabletRelease` is not always going to be the last point to draw.  If you draw quickly enough, by the time the release event is processed, the actual point may be different from where they intended the stroke to end. This part of the code is finalizing the stroke by adding the release postion to the end of the stroke so it's drawing the line to where the release actually happened.

Normally they are pretty close so you won't notice it.  I had to draw fast enough and zoom in to see little tails being drawn at the end of the stroke. Lazy Nezumi makes this issue a lot more obvious since the pointer is not going to be the same as what is being drawn.

I corrected the problem by tracking the last screen position in `onMove()`, and using that to finalize the tracking information as follows in `onRelease()`:

```
---------- /toonz/sources/toonz/sceneviewer.h ----------
@@ -73,6 +73,7 @@ class SceneViewer final : public GLWidgetForHighDpi,

   double m_pressure;
   QPointF m_lastMousePos;
+  TPointD m_lastScreenPos;
   QPointF m_pos;
   Qt::MouseButton m_mouseButton;
   bool m_foregroundDrawing;

---------- /toonz/sources/toonz/sceneviewerevents.cpp ----------
@@ -521,6 +521,8 @@ void SceneViewer::onMove(const TMouseEvent &event) {
     if (m_locator) m_locator->onChangeViewAff(worldPos);
     TPointD pos = getInputManager()->worldToTool() * worldPos;

+       m_lastScreenPos = screenPos;
+
     // qDebug() << "mouseMoveEvent. "  << (m_tabletEvent?"tablet":"mouse")
     //         << " pressure=" << m_pressure << " mouseButton=" << m_mouseButton
     //         << " buttonClicked=" << m_buttonClicked;
@@ -796,7 +798,7 @@ void SceneViewer::onRelease(const TMouseEvent &event) {
     if (m_mouseButton == Qt::LeftButton || m_tabletState == Released) {
       if (!m_toolSwitched) {
         getInputManager()->trackEvent(
-          0, 0, screenPos,
+          0, 0, m_lastScreenPos,
           (event.m_isTablet ? &event.m_pressure : NULL),
           NULL, true, TToolTimer::ticks() );
         getInputManager()->processTracks();
```

I think, for the most part, the release position is or should be ignored though we'd need to check all the tools to make certain it doesn't need exact release position.